installer template: add 'recover from etcd quorum loss' test

vrutkovs opened the pull request:
Rework of #3495 - run etcd quorum loss as a dedicated test in installer template.

This adds a new optional test `e2e-etcd-quorum-loss` which simulates the quorum loss. At this point its expected to fail as a single etcd node doesn't get restored yet.

This PR uses a forked version of scripts served by MCO. This would be fixed in a separate PR so that we could work in parallel on various small tasks related to DR

TODO:
* [x] Rebase when https://github.com/openshift/release/pull/3595 lands

wking commented on ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml:
```diff
@@ -159,6 +159,87 @@ objects:
+          # Verify etcdctl member list returns 3 members
+          # Verify no alerts are firing
+
+          openshift-tests run-upgrade "${TEST_SUITE}" --to-image "${RELEASE_IMAGE_LATEST}" \
```
No need to embed this.  The caller can use:

```yaml
commands: recover-from-etcd-quorum-loss && TEST_SUITE=openshift/conformance/parallel run-tests
```

Or:

```yaml
commands: |
  recover-from-etcd-quorum-loss
  TEST_SUITE=openshift/conformance/parallel run-tests
```

vrutkovs commented:
```
Failing tests: [Feature:Platform][Smoke] Managed cluster should ensure control plane operators do not make themselves unevictable [Suite:openshift/conformance/parallel] [Feature:Platform][Smoke] Managed cluster should ensure pods use images from our release image with proper ImagePullPolicy [Suite:openshift/conformance/parallel]
```
The latter is remaining ssh-bastion, fixed in latest commit

hexfusion commented on ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml:
```diff
@@ -159,6 +163,250 @@ objects:
+          echo "Run etcd-signer"
+          SURVIVING_MASTER_NODE_SHORT=${SURVIVING_MASTER_NODE%%.*}
+          curl -O https://raw.githubusercontent.com/hexfusion/openshift-recovery/master/manifests/kube-etcd-cert-signer.yaml.template
+          sed "s;__MASTER_HOSTNAME__;${SURVIVING_MASTER_NODE_SHORT};g" kube-etcd-cert-signer.yaml.template > kube-etcd-cert-signer.yaml
```
Lets update this to use file from disk now taht MCO has merged https://github.com/openshift/machine-config-operator/blob/master/templates/master/00-master/_base/files/usr-local-share-openshift-recovery-template-kube-etcd-cert-signer-yaml-template.yaml#L3

hexfusion replied:
Then use https://github.com/openshift/machine-config-operator/blob/master/templates/master/00-master/_base/files/usr-local-bin-tokenize-signer-sh.yaml

vrutkovs replied:
Right, I'd rather take care of that in a separate PR so that we could fix few things in parallel

wking commented on ci-operator/config/openshift/installer/openshift-installer-master.yaml:
```diff
@@ -152,3 +152,11 @@ tests:
     TEST_SUITE=openshift/conformance/parallel run-tests
   openshift_installer:
     cluster_profile: aws
+
```
nit: I don't see a need to split this out from the rest of the the tests defined above.  Can we drop this empty line and alphabetize the new entry into the block above?

vrutkovs replied:
Fixed in 603449306

wking commented on ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml:
```diff
@@ -342,6 +356,240 @@ objects:
+          mkdir -p ~/.ssh || true
+          cp "${KUBE_SSH_KEY_PATH}" ~/.ssh/id_rsa
+          chmod 0600 ~/.ssh/id_rsa
+          echo "${USER:-default}:x:$(id -u):$(id -g):Default User:$HOME:/sbin/nologin" >> /etc/passwd
```
Do we want guards around this?  Similar logic in other places [has guards][1].

[1]: https://github.com/openshift/release/blob/6ca4a9371305b26407c90c2fe82a98c83b93591a/ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml#L533-L534

vrutkovs replied:
Fixed in 8ddaca4e4

wking commented on ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml:
```diff
@@ -342,6 +356,240 @@ objects:
+          echo "${USER:-default}:x:$(id -u):$(id -g):Default User:$HOME:/sbin/nologin" >> /etc/passwd
+
+          echo "Setting up ssh bastion"
+          curl https://raw.githubusercontent.com/eparis/ssh-bastion/master/deploy/deploy.sh | bash
```
It would be nice to share code instead of duplicating [this existing logic][1] here.  But I guess this is all going to go away a few minutes after this PR lands? ;)

[1]: https://github.com/openshift/release/blob/6ca4a9371305b26407c90c2fe82a98c83b93591a/ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml#L197

vrutkovs replied:
Fixed in 1484b5006

wking commented on ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml:
```diff
@@ -342,6 +356,240 @@ objects:
+                  $@
+          }
+
+          # Machine API won't let the user to destroy the node which runs the controller
```
Is there a bug for this?  Seems like a bug.  Doesn't mean we can't work around it here, but I don't want to forget about fixing it ;).

vrutkovs replied:
https://bugzilla.redhat.com/show_bug.cgi?id=1700931 - this is a limitation of Machine API

wking commented on ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml:
```diff
@@ -342,6 +356,240 @@ objects:
+          done
+
+          echo "Confirm meltdown"
+          sleep 30
```
What are we waiting for here?  This seems brittle.

vrutkovs replied:
Machine API won't wait for machine to be deleted during `oc delete machine` call, so API is available for a few seconds after 2 masters were removed.

This pause is necessary to confirm API is no longer responding

wking commented on ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml:
```diff
@@ -342,6 +356,240 @@ objects:
+          exit 1
+        fi
+
+        PREFIX="https://raw.githubusercontent.com/vrutkovs/openshift-recovery/unify-fixes"
```
floating downloads :sob: :wink:

vrutkovs replied:
Its a temporary measure while we debug MCO scripts.

This is fixed in #3842 which is the same PR + MCO scripts

wking commented on ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml:
```diff
@@ -342,6 +356,240 @@ objects:
+          SURVIVING_MASTER_AZ_INDEX=$(grep -oP 'availabilityZone: .*(\K.)' /tmp/machine.template)
+
+          MASTER_INDEX=0
+          AZ_LIST=("a" "b" "c")
```
This seems very brittle (e.g. we currently only use [two zones][1]).  Can we pull it from `oc get -o 'jsonpath={...}' nodes` output or some such instead of hard-coding it?  Or is this enumeration not really about availability zones?

[1]: https://github.com/openshift/release/blob/6ca4a9371305b26407c90c2fe82a98c83b93591a/ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml#L426-L436

vrutkovs replied:
Fixed in d50d50e33 - a simple loop is used to create two missing masters. Both would be put in the same AZ as the master.

Previously I was attempting to put new masters in other AZs. AZ list had to be kept in sync with install settings, so I dropped it.

wking commented on ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml:
```diff
@@ -342,6 +356,240 @@ objects:
+          # aws cli magic
+          easy_install --user 'setuptools==12.0'
+          easy_install --user boto3
+          cat > /tmp/update_route_53.py <<-PYTHON_EOF
```
You're not actually using the tab-indents that `<<-` allows, are you?  Might as well stick with `<<PYTHON_EOF`.

vrutkovs replied:
Right, fixed in a60bc18b4

wking commented on ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml:
```diff
@@ -342,6 +356,240 @@ objects:
+          bastion_ssh core@${SURVIVING_MASTER_NODE} "sudo -i /usr/local/bin/etcd-snapshot-restore.sh"
+
+          echo "Wait for API server to come up"
+          retry 30 oc get nodes
```
Not part of the test framework that I can review, but if your etcd cluster is down to one node (because you killed two of three control-plane nodes), then how is the remaining etcd still functioning?  I'd have expected it to be freaking out about having lost quorum and refusing to take possibly-split-brained actions.  If there's some quick explanation for how this works, I'm very curious.  If the explanation is longer, we should probably skip it to avoid distracting from the test script itself.

vrutkovs replied:
>if your etcd cluster is down to one node (because you killed two of three control-plane nodes), then how is the remaining etcd still functioning?

`etcd-snapshot-restore.sh` without params would restore this etcd forming a one node cluster - https://github.com/openshift/machine-config-operator/blob/master/templates/master/00-master/_base/files/usr-local-bin-openshift-recovery-tools-sh.yaml#L141-L168.

Added a bit better comment for this in b99911295

wking commented on ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml:
```diff
@@ -342,6 +356,240 @@ objects:
+
+          #echo "Verify kubernetes svc points to 3 pods"
+
+          # Verify no alerts are firing
```
Probably drop these commented-out placeholders and file issues if you need a way to track future plans.

vrutkovs replied:
Fixed in 143155551

vrutkovs commented:
Fixed all outstanding issues, rebased on master. 

This would eventually be substituted by #3842, changes to use MCO are contained in https://github.com/openshift/release/pull/3842/commits/ad5a69f2fe8de3660f0280c13aec2086905a0329

vrutkovs commented:
```
fail [github.com/openshift/origin/test/extended/operators/cluster.go:118]: Expected
    <[]string | len:4, cap:4>: [
        "Pod openshift-apiserver-operator/openshift-apiserver-operator-56c559db6f-xfmhm is not healthy: container openshift-apiserver-operator has restarted more than 5 times",
        "Pod openshift-controller-manager-operator/openshift-controller-manager-operator-85cc57f9d7-qvgnm is not healthy: container operator has restarted more than 5 times",
        "Pod openshift-kube-apiserver-operator/kube-apiserver-operator-84c9f5b76c-rklfj is not healthy: container kube-apiserver-operator has restarted more than 5 times",
        "Pod openshift-kube-scheduler-operator/openshift-kube-scheduler-operator-6c677b65cc-rmw7r is not healthy: container kube-scheduler-operator-container has restarted more than 5 times",
    ]
to be empty
```

/test pj-rehearse

vrutkovs commented:
Superceded by #3842
